### DISCUSSION TITLE
To make startup of pupcloud easier we can use a config file

### DISCUSSION THREAD

**Discussion by marioscube** [proofrock/pupcloud#26] [Ideas]
When I first used pupcloud I started it with 

pupcloud -r /home/user/pupcloud

That is easy and relatively quick to type.

After some experimenting with pupcloud my startup line evaluated to 

pupcloud -r /home/user/Share/pubcloud -H 1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef --max-upload-size 128 -E --share-profile John:S3cr3t --share-profile James:S5cr5t --share-prefix 'http://domain.com:12345/pupshare' --follow-symlinks

This is a bit much to type .............

ENHANCEMENT: wouldn't it be nice to ha a config file (pupcloud.conf) to start pupcloud.

pupcloud might be started with: pupcloud -c pupcloud.conf
option -c is config.file

**Comment by proofrock** [comment]
I feel you're right 😬; it was a conscious design decision to try and do without any kind of database, dotfile _or config file_... right or not, I wanted to play within the constraint and see what I can still do. 

To address your issue: from version 0.8.0 pupcloud [there is support for env variables](https://germ.gitbook.io/pupcloud/guides/running-pupcloud#configure-by-env-vars). They can be used to write a "rich" shell script that looks like a config file, maybe? Something like:

```bash
export PUP_ROOT=/home/user/Share/pubcloud
export PUP_PWD_HASH=1234567890abcde...
export PUP_MAX_UPLOAD_SIZE=128
export PUP_ALLOW_EDITS=1
export PUP_SHARE_PROFILES=John:S3cr3t,James:S5cr5t
export PUP_SHARE_PREFIX=http://domain.com:12345/pupshare
export PUP_FOLLOW_SYMLINKS=1
./pupcloud
```
You need to `export` them, to make them visible from a child process. Also putting in the same command as pupcloud works, but to format it you must use `\`:

```bash
PUP_ROOT=/home/user/Share/pubcloud \
PUP_PWD_HASH=1234567890abcde... \
PUP_MAX_UPLOAD_SIZE=128 \
PUP_ALLOW_EDITS=1 \
PUP_SHARE_PROFILES=John:S3cr3t,James:S5cr5t \
PUP_SHARE_PREFIX=http://domain.com:12345/pupshare \
PUP_FOLLOW_SYMLINKS=1 \
./pupcloud
```

**Comment by proofrock** [comment]
I wrote [some docs](https://germ.gitbook.io/pupcloud/guides/running-pupcloud#write-a-config-file) reflecting this discussion, thanks.

**Comment by marioscube** [comment]
I thought it would be difficult using env variables, but with your documentation it was easy.

I have already made a bash file to use with the next release of pupcloud.

Only the user that sets the env variables can change them right? (except for root obviously)

**Comment by proofrock** [comment]
Good to hear! Nice.

> Only the user that sets the env variables can change them right? (except for root obviously)

I am not a sysadmin, so I don't really know for sure; what I can tell you is that, in pupcloud's code, the env vars get copied at startup into local variables, so even if they changed value after startup it wouldn't matter. 

Also, the second form I mentioned directly "pipe" the vars in the command, so it's much more difficult (impossible?) to change them. It's an atomic operation, so to speak.

**Comment by marioscube** [comment]
export PUP_ROOT=/path        Works

PUP_ROOT=/path \                   Does not work for me
Pupcloud v0.8.0RC1 (c) 2022 Germano Rizzo
ERROR: you must specify a root dir (-r)

**Comment by proofrock** [comment]
The sense of the second "form" is that this works:

```bash
PUP_ROOT=/path ./pupcloud
```
It basically defines a variable for the command that follows it. 

In bash at least (I don't know other shells) the character `\` escapes the newline, so the following is the same:

```bash
PUP_ROOT=/path \
./pupcloud
```

Take care to avoid spaces after the `\`, and to avoid blank lines in between, also. I am not an expert in shell scripting, but I never had any problem with this, it's also quite common to write long commands (ex. docker) like this.

**Comment by marioscube** [comment]
I typed it all again and now it works. 

Must have been that rogue space you wrote about.

thank you.
